eth/protocols/snap/sync_test.go

```diff
@@ -280,48 +286,40 @@ func defaultCodeRequestHandler(t *testPeer, id uint64, hashes []common.Hash, max
 	return nil
 }
 
-func createStorageRequestResponse(t *testPeer, root common.Hash, accounts []common.Hash, bOrigin, bLimit []byte, max uint64) (hashes [][]common.Hash, slots [][][]byte, proofs [][]byte) {
-	var (
-		size  uint64
-		limit = common.HexToHash("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
-	)
-	if len(bLimit) > 0 {
-		limit = common.BytesToHash(bLimit)
-	}
-	var origin common.Hash
-	if len(bOrigin) > 0 {
-		origin = common.BytesToHash(bOrigin)
-	}
-
-	var limitExceeded bool
-	var incomplete bool
+func createStorageRequestResponse(t *testPeer, root common.Hash, accounts []common.Hash, origin, limit []byte, max uint64) (hashes [][]common.Hash, slots [][][]byte, proofs [][]byte) {
+	var size uint64
 	for _, account := range accounts {
-
-		var keys []common.Hash
-		var vals [][]byte
+		// The first account might start from a different origin and end sooner
+		var originHash common.Hash
+		if len(origin) > 0 {
+			originHash = common.BytesToHash(origin)
+		}
+		var limitHash = common.HexToHash("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+		if len(limit) > 0 {
+			limitHash = common.BytesToHash(limit)
+		}
+		var (
+			keys []common.Hash
+			vals [][]byte
+		)
 		for _, entry := range t.storageValues[account] {
-			if limitExceeded {
-				incomplete = true
-				break
-			}
-			if bytes.Compare(entry.k, origin[:]) < 0 {
-				incomplete = true
+			if bytes.Compare(entry.k, originHash[:]) < 0 {
 				continue
 			}
 			keys = append(keys, common.BytesToHash(entry.k))
 			vals = append(vals, entry.v)
 			size += uint64(32 + len(entry.v))
-			if bytes.Compare(entry.k, limit[:]) >= 0 {
-				limitExceeded = true
-			}
-			if size > max {
-				limitExceeded = true
+			if bytes.Compare(entry.k, limitHash[:]) >= 0 || size > max {
+				break
 			}
 		}
 		hashes = append(hashes, keys)
 		slots = append(slots, vals)
 
-		if incomplete {
+		// Generate the Merkle proofs for the first and last storage slot, but
+		// only if the response was capped. If the entire storage trie included
+		// in the response, no need for any proofs.
+		if originHash != (common.Hash{}) || size >= max {
```

Comment:
There exist a cornecase where `size > max` _and_ the response is complete. Hence why I set `incomplete=true` when `limitExceeded=true` _and_ I knew there were more items that could be sent.

Reply:
Yes, it's possible. But I think the downside is OK, we only generate the "unnecessary" proof for this special case in the tests. I think it's acceptable for the simpler code perhaps?

But I'm fine to add it back.

Reply:
The "unnecessary" proof is also handled properly. So I think we can keep it :P

Reply:
Fixed